src/fondant/core/manifest.py

```diff
@@ -238,11 +238,76 @@ def remove_field(self, name: str) -> None:
 
         del self._specification["fields"][name]
 
+    @staticmethod
+    def _handle_custom_produces(
+        evolved_manifest: "Manifest",
+        produces: t.Optional[t.Union[t.Dict[str, str], t.Dict[str, Field]]],
+        component_spec: ComponentSpec,
+    ) -> t.Mapping[str, Field]:
+        handled_produces: t.Dict[str, Field] = {}
+        is_produces_generic = component_spec.is_produces_generic
+
+        # Add component spec produces
+        if component_spec.produces:
+            for column_name, field in component_spec.produces.items():
+                handled_produces[column_name] = field
+
+        # Add custom produces defined in the component operation
+        if produces:
+            for column_name, mapping_name_or_type in produces.items():
+                # String mapping name
+                if isinstance(mapping_name_or_type, str):
+                    try:
+                        field = handled_produces[column_name]
+                    except KeyError:
+                        msg = (
+                            f"Field {column_name} not found in component spec produces: "
+                            f"{component_spec.produces}"
+                        )
+                        raise InvalidPipelineDefinition(msg)
+
+                    field.name = mapping_name_or_type
+                    handled_produces[mapping_name_or_type] = field
+                    # Remove the original field if it was already in the manifest
+                    if column_name in evolved_manifest.fields:
```

Comment:
I would prefer to not modify the `evolved_manifest` in this function. I would let it resolve the `produces` section based on the component spec and the `produces` argument of the ComponentOp. And then use that resolved `produces` section to evolve the manifest.